test/test_nn.py

```diff
@@ -11641,6 +11695,178 @@ def test_embedding_padding_idx(self, device, dtype):
+    # backward functions with padding_idx, given a 2D indices input. Compare
+    # against torch.nn.functional.embedding followed by a reduction.
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float32, torch.float64)
+    def test_embedding_bag_2D_padding_idx(self, device, dtype):
+        # Use a Python implementation of embedding_bag with padding_idx support
+        # to check torch.nn.functional.embedding_bag correctness
+        def embedding_bag_check(indices, weights, mode, sparse, padding_idx):
+            assert padding_idx is not None
+            embedding = torch.nn.functional.embedding(
+                indices,
+                weights,
+                padding_idx=padding_idx,
+                sparse=sparse)
+
+            reduction_dim = indices.dim() - 1
+
+            if mode == 'sum' or mode == 'mean':
+                per_sample_weights = indices.ne(padding_idx).to(dtype).unsqueeze(-1)
+                res = embedding.mul(per_sample_weights).sum(dim=reduction_dim)
+
+                if mode == 'mean':
+                    res = res.div(per_sample_weights.sum(dim=reduction_dim))
+            elif mode == 'max':
+                res = embedding.masked_fill(
+                    indices.unsqueeze(-1) == padding_idx, -float('inf')
+                ).amax(dim=reduction_dim)
+            else:
+                raise RuntimeError(f"mode '{mode}' is not available")
+
+            return res
+
+        num_features = 3
+        num_words = 10
+        indices_dim0 = 10
+        indices_dim1 = 12
+
+        for mode in ['max', 'mean', 'sum']:
+            for sparse in [False]:
+                indices = torch.randint(0, num_words, (indices_dim0, indices_dim1), device=device)
+                for padding_idx in list(set(indices.flatten(0, -1).tolist())):
+                    weights = torch.randn(num_words, num_features, dtype=dtype, device=device, requires_grad=True)
+                    weights_check = weights.clone().detach().requires_grad_(True)
+
+                    msg = f"mode: '{mode}', sparse: {sparse}, padding_idx: {padding_idx}"
+
+                    # Check forward with a Python implementation of padding_idx embedding_bag
+                    bag_check = embedding_bag_check(
+                        indices,
+                        weights_check,
+                        mode,
+                        sparse,
+                        padding_idx)
+                    bag = torch.nn.functional.embedding_bag(
+                        indices,
+                        weights,
+                        padding_idx=padding_idx,
+                        mode=mode)
+
+                    self.assertEqual(bag, bag_check, msg=msg)
+
+                    bag_check.sum().backward()
+                    grad_check = weights_check.grad
+
+                    bag.sum().backward()
+                    grad = weights.grad
+
+                    self.assertEqual(grad, grad_check, msg=msg)
+
```

Comment:
What do you think about adding tests for the all-padding input cases for max, sum, and mean? Or do you think the empty bag cases cover this well enough?

Reply:
Good idea, I've added allpad cases to both 1D and 2D input tests